File: lib/writelog.c
#include "common.h"
#include "composer.h"
#include "config.h"
#include "getinfo.h"
#include "hprintf.h"
#include "writer.h"

#include <git2/commit.h>
#include <git2/revwalk.h>
#include <limits.h>
#include <unistd.h>

static void writelogline(FILE* fp, git_commit* commit, const struct commitinfo* ci) {
	char                 oid[GIT_OID_SHA1_HEXSIZE + 1];
	const git_signature* author  = git_commit_author(commit);
	const char*          summary = git_commit_summary(commit);

	git_oid_tostr(oid, sizeof(oid), git_commit_id(commit));

	fputs("<tr><td>", fp);
	if (author)
		hprintf(fp, "%t", &author->when);
	fputs("</td><td>", fp);
	if (summary) {
		hprintf(fp, "<a href=\"../commit/%s.html\">%y</a>", oid, summary);
	}
	fputs("</td><td>", fp);
	if (author)
		hprintf(fp, "%y", author->name);
	fputs("</td><td class=\"num\" align=\"right\">", fp);
	fprintf(fp, "%zu", ci->ndeltas);
	fputs("</td><td class=\"num\" align=\"right\">", fp);
	fprintf(fp, "+%zu", ci->addcount);
	fputs("</td><td class=\"num\" align=\"right\">", fp);
	fprintf(fp, "-%zu", ci->delcount);
	fputs("</td></tr>\n", fp);
}

static void writelogcommit(FILE* fp, FILE* json, FILE* atom, const struct repoinfo* info, int index,
                           git_oid* id) {
	struct commitinfo ci;
	git_commit*       commit;
	git_tree*         tree;
	char              path[PATH_MAX], oid[GIT_OID_SHA1_HEXSIZE + 1];
	int               cachedcommit;

	/* Lookup the current commit */
	if (git_commit_lookup(&commit, info->repo, id)) {
		hprintf(stderr, "error: unable to lookup commit: %gw\n");
		return;
	}

	/* Get the tree for the current commit */
	if (git_commit_tree(&tree, commit)) {
		hprintf(stderr, "error: unable to get tree for commit: %gw\n");
		git_commit_free(commit);
		return;
	}

	git_oid_tostr(oid, sizeof(oid), id);
	snprintf(path, sizeof(path), "commit/%s.html", oid);

	cachedcommit = !force && !access(path, F_OK);
	if (getdiff(&ci, info, commit, cachedcommit) == -1)
		return;

	if (!cachedcommit) {
		FILE* commitfile = efopen("w", "%s", path);
		writecommit(commitfile, info, commit, &ci, index == maxcommits);
		fclose(commitfile);
	}
	if (fp)
		writelogline(fp, commit, &ci);
	if (atom)
		writeatomcommit(atom, commit, NULL);
	if (json)
		writejsoncommit(json, commit, index == 0);

	freediff(&ci);
	git_commit_free(commit);
	git_tree_free(tree);
}

int writelog(FILE* fp, FILE* atom, FILE* json, const struct repoinfo* info,
             struct referenceinfo* refinfo) {
	git_revwalk* w = NULL;
	git_oid      id;
	ssize_t      ncommits = 0, arsize;
	FILE*        arfp;
	const char*  unit;

	emkdirf("commit");
	emkdirf("%s", refinfo->refname);

	if (fp) {
		writeheader(fp, info, 1, 1, info->name, "%s", refinfo->refname);

		fputs("<h2>Archives</h2>", fp);
		fputs("<table><thead>\n<tr><td class=\"expand\">Name</td>"
		      "<td class=\"num\">Size</td></tr>\n</thead><tbody>\n",
		      fp);
	}

	FORMASK(type, archivetypes) {
		arfp   = efopen("w+", "%s/%s.%s", refinfo->refname, refinfo->refname, archiveexts[type]);
		arsize = writearchive(arfp, info, type, refinfo);
		fclose(arfp);
		unit = splitunit(&arsize);
		if (fp)
			fprintf(fp, "<tr><td><a href=\"%s.%s\">%s.%s</a></td><td>%zd%s</td></tr>",
			        refinfo->refname, archiveexts[type], refinfo->refname, archiveexts[type],
			        arsize, unit);
	}

	if (fp) {
		fputs("</tbody></table>", fp);

		writeshortlog(fp, info, refinfo->commit);

		fprintf(fp, "<h2>Commits of %s</h2>", refinfo->refname);

		fprintf(fp, "<table id=\"log\"><thead>\n<tr><td>Date</td>"
		            "<td class=\"expand\">Commit message</td>"
		            "<td>Author</td><td class=\"num\" align=\"right\">Files</td>"
		            "<td class=\"num\" align=\"right\">+</td>"
		            "<td class=\"num\" align=\"right\">-</td></tr>\n</thead><tbody>");
	}
	if (json)
		fprintf(json, "{\"commits\":{");

	if (atom)
		writeatomheader(atom, info);

	/* Create a revwalk to iterate through the commits */
	if (git_revwalk_new(&w, info->repo)) {
		hprintf(stderr, "error: unable to initialize revwalk: %gw\n");
		return -1;
	}
	git_revwalk_push(w, git_commit_id(refinfo->commit));

	/* Iterate through the commits */
	while (!git_revwalk_next(&id, w))
		ncommits++;

	git_revwalk_reset(w);
	git_revwalk_push(w, git_commit_id(refinfo->commit));

	/* Iterate through the commits */
	ssize_t indx = 0;
	while ((maxcommits <= 0 || indx < maxcommits) && !git_revwalk_next(&id, w)) {
		writelogcommit(fp, json, atom, info, indx, &id);
		indx++;

		printprogress(indx, ncommits, "write log:   %-20s", refinfo->refname);
	}
	if (!verbose && !quiet)
		fputc('\n', stdout);

	git_revwalk_free(w);

	if (fp) {
		if (maxcommits > 0 && ncommits > maxcommits) {
			if (ncommits - maxcommits == 1)
				fprintf(fp, "<tr><td></td><td colspan=\"5\">1 commit left out...</td></tr>\n");
			else
				fprintf(fp, "<tr><td></td><td colspan=\"5\">%lld commits left out...</td></tr>\n",
				        ncommits - maxcommits);
		}

		fputs("</tbody></table>", fp);
		writefooter(fp);
	}

	if (json)
		fprintf(json, "}}");

	if (atom)
		writeatomfooter(atom);

	return 0;
}

File: internal/writer/writelog.go
package writer

import (
	"errors"
	"fmt"
	"html"
	"io"
	"os"
	"path"
	"strings"

	"github.com/friedelschoen/gitin-go"
	"github.com/friedelschoen/gitin-go/internal/common"
	"github.com/friedelschoen/gitin-go/internal/wrapper"
	git "github.com/jeffwelling/git2go/v37"
)

func writelogline(fp io.Writer, commit *git.Commit, ci *wrapper.CommitInfo) {
	author := commit.Author()
	summary := commit.Summary()

	oid := commit.Id().String()

	fmt.Fprint(fp, "<tr><td>")
	if author != nil {
		fmt.Fprint(fp, rfc3339(author.When))
	}
	fmt.Fprint(fp, "</td><td>")
	if summary != "" {
		fmt.Fprintf(fp, "<a href=\"../commit/%s.html\">%s</a>", oid, html.EscapeString(summary))
	}
	fmt.Fprintf(fp, "</td><td>")
	if author != nil {
		fmt.Fprintf(fp, "%s", html.EscapeString(author.Name))
	}
	fmt.Fprintf(fp, "</td><td class=\"num\" align=\"right\">")
	fmt.Fprintf(fp, "%d", len(ci.Deltas))
	fmt.Fprintf(fp, "</td><td class=\"num\" align=\"right\">")
	fmt.Fprintf(fp, "+%d", ci.Addcount)
	fmt.Fprintf(fp, "</td><td class=\"num\" align=\"right\">")
	fmt.Fprintf(fp, "-%d", ci.Delcount)
	fmt.Fprint(fp, "</td></tr>\n")
}

func FileExist(pat string) bool {
	_, err := os.Stat(pat)
	return !errors.Is(err, os.ErrNotExist)
}

type logstate struct {
	fp   io.Writer
	json []commitJSON
	atom []atomEntry
}

func writelogcommit(s *logstate, info *wrapper.RepoInfo, index int, commit *git.Commit) error {
	pat := path.Join("commit", commit.Id().String()+".html")
	cachedcommit := !common.Force && FileExist(pat)
	ci, err := wrapper.GetDiff(info, commit, cachedcommit)
	if err != nil {
		return err
	}

	if !cachedcommit {
		commitfile, err := os.Create(pat)
		if err != nil {
			return err
		}
		defer commitfile.Close()
		if err := writecommit(commitfile, info, commit, &ci, index == gitin.Config.Maxcommits); err != nil {
			return err
		}
	}
	if s.fp != nil {
		writelogline(s.fp, commit, &ci)
	}
	s.atom = append(s.atom, makeEntry(commit, ""))
	s.json = append(s.json, toJSONCommit(commit))
	return nil
}

func writelog(fp io.Writer, atom io.Writer, json io.Writer, info *wrapper.RepoInfo, refinfo *wrapper.ReferenceInfo) error {

	os.Mkdir("commit", 0777)
	os.Mkdir(refinfo.Refname, 0777)

	if fp != nil {
		writeheader(fp, info, 1, true, info.Name, refinfo.Refname)

		fmt.Fprintf(fp, "<h2>Archives</h2>")
		fmt.Fprintf(fp, "<table><thead>\n<tr><td class=\"expand\">Name</td>"+
			"<td class=\"num\">Size</td></tr>\n</thead><tbody>\n")
	}

	for _, ext := range gitin.Config.Archives {
		ext = strings.TrimPrefix(ext, ".") /* .tar.xz -> tar.xz, .zip -> zip */
		arfp, err := os.Create(path.Join(refinfo.Refname, refinfo.Refname+"."+ext))
		if err != nil {
			return err
		}
		defer arfp.Close()
		arsize, err := writearchive(arfp, info, ext, refinfo)
		arsize, unit := common.Splitunit(arsize)
		if fp != nil {
			fmt.Fprintf(fp, "<tr><td><a href=\"%s.%s\">%s.%s</a></td><td>%d%s</td></tr>", refinfo.Refname, ext, refinfo.Refname, ext, arsize, unit)
		}
	}

	if fp != nil {
		fmt.Fprintf(fp, "</tbody></table>")

		if err := writeshortlog(fp, info, refinfo.Commit); err != nil {
			return err
		}

		fmt.Fprintf(fp, "<h2>Commits of %s</h2>", refinfo.Refname)

		fmt.Fprintf(fp, "<table id=\"log\"><thead>\n<tr><td>Date</td>"+
			"<td class=\"expand\">Commit message</td>"+
			"<td>Author</td><td class=\"num\" align=\"right\">Files</td>"+
			"<td class=\"num\" align=\"right\">+</td>"+
			"<td class=\"num\" align=\"right\">-</td></tr>\n</thead><tbody>")
	}

	/* Create a revwalk to iterate through the commits */
	w, err := info.Repo.Walk()
	if err != nil {
		return fmt.Errorf("unable to initialize revwalk: %w", err)
	}

	if err := w.Push(refinfo.Commit.Id()); err != nil {
		return err
	}
	ncommits := 0
	_ = w.Iterate(func(commit *git.Commit) bool {
		ncommits++
		return true
	})
	w.Reset()
	if err := w.Push(refinfo.Commit.Id()); err != nil {
		return err
	}

	msg := fmt.Sprintf("write log:   %s", refinfo.Refname)
	defer common.Printer.Done(msg)

	/* Iterate through the commits */
	var indx int = 0
	state := logstate{fp: fp}
	_ = w.Iterate(func(commit *git.Commit) bool {
		if gitin.Config.Maxcommits > 0 && indx >= gitin.Config.Maxcommits {
			return false
		}
		if err = writelogcommit(&state, info, indx, commit); err != nil {
			return false
		}
		indx++

		common.Printer.Progress(msg, indx, ncommits)
		return true
	})
	if err != nil {
		return err
	}
	if !common.Verbose && !common.Quiet {
		fmt.Println()
	}

	if fp != nil {
		if gitin.Config.Maxcommits > 0 && ncommits > gitin.Config.Maxcommits {
			var plural string
			if ncommits-gitin.Config.Maxcommits != 1 {
				plural = "s"
			}
			fmt.Fprintf(fp, "<tr><td></td><td colspan=\"5\">%d%s commits left out...</td></tr>\n", ncommits-gitin.Config.Maxcommits, plural)
		}

		fmt.Fprintf(fp, "</tbody></table>")
		writefooter(fp)
	}
	if json != nil {
		if err := writejsoncommits(json, state.json); err != nil {
			return err
		}
	}
	if atom != nil {
		if err := writeatomfeed(atom, info, state.atom); err != nil {
			return err
		}
	}
	return nil
}
